relay-cardinality/src/cardinality.lua

```diff
@@ -0,0 +1,174 @@
+--
+--
+-- ``KEYS``: A list of cardinality sets, the first key 'working set' is used to check cardinality,
+-- hashes will be updated in all passed keys.
+--
+-- ``ARGV``:
+--  * [number] Max cardinality.
+--  * [number] Set expiry.
+--  * [string...] List of hashes.
+--
+--  Returns a table, the first element in the table contains the new set cardinality and for
+--  every passed hash (in order) the table contains `0` if the hash was rejected and `1` if it was accepted.
+--
+--
+-- For example to check the cardinality limit of 3 with the hashes `1, 2, 3, 4, 5`
+-- in a sliding window of 1 hour with a granularity of 2 (sets: `0` and `1800`)
+-- send the following values:
+--
+--      KEYS: { "prefix-scope-0", "prefix-scope-1800" }
+--      ARGV: {
+--          3,             -- Limit
+--          3600,          -- Window size / Expiry
+--          1, 2, 3, 4, 5  -- Hashes
+--      }
+--
+-- The script returns:
+--
+--      {
+--          3, -- new cardinality
+--          1, -- accepted: 1
+--          1, -- accepted: 2
+--          1, -- accepted: 3
+--          0, -- rejected: 4
+--          0, -- rejected: 5
+--      }
+--
+-- Redis state after execution:
+--
+--      prefix-scope-0:    {1, 2, 3} | Expires: now + 3600
+--      prefix-scope-1800: {1, 2, 3} | Expires: now + 3600
+--
+--
+-- The script applies the following logic for every hash passed as an argument:
+--  * if the cardinality has not been reached yet, the hash is added to all sets
+--    and the item is marked as accepted
+--  * otherwise it only marks the hash as accepted when the hash was already seen before,
+--    this is done by checking wether the hash is contained in the 'working set'
+--
+-- Afterwards if any item was added to the 'working set' the expiry of all sets is bumped
+-- with the passed expiry.
+
+local function batches(n, batch_size, offset, max_items)
+    local i = 0
+    if max_items then
+        n = math.min(n, max_items + offset)
+    end
+
+    return function()
+        local from = i * batch_size + 1 + offset
+        i = i + 1
+        if from <= n then
+            local to = math.min(from + batch_size - 1, n)
+
+            return from, to
+        end
+    end
+end
+
+
+local ACCEPTED = true
+local REJECTED = false
+local HASHES_OFFSET = 2 -- arg1: max_cardinality, arg2: expiry
+
+local working_set = KEYS[1]
+local max_cardinality = tonumber(ARGV[1])
+local expire = tonumber(ARGV[2])
+local num_hashes = #ARGV - HASHES_OFFSET
+
+local results = {
+    0, -- total cardinality
+}
+
+-- Adds elements from the table `t`, starting with index `offset` and a maximum of `max` elements
+-- to all sets in `KEYS`.
+--
+-- Returns the total amount of values added to the 'working set'.
+local function sadd(t, offset, max)
+    local added = 0;
+
+    for i = 1, #KEYS do
+        local is_working_set = i == 1
+
+        for from, to in batches(#t, 7000, offset, max) do
```

Comment:
I think we should batch exclusively on the Relay side. That would simplify this script, and we need some sort of batching there anyway.

Reply:
I like batching in the lua script, makes the relay side easier to read and we have a guarantee that this never fails in lua (one less error to handle and invariant).

The lua implementation would get slightly easier, but we would still need a slicing function like `batches` (to slice `offset` and `max`), which now is just a minimal overhead to the `batches` function (just a change of the max value and the offset).

So overall imo we don't gain much by moving it to relay.

Reply:
I am not sure how Redis and the network behaves here, in my local tests it didn't make a difference but there is also no latency and Redis does not do anything else (e.g. rate limiting etc.)

Reply:
Would we batch and use a pipeline, I feel like that would make no difference then?